hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java

```diff
@@ -42,22 +67,202 @@ public TokenAssociateToAccountHandler() {
+            validateTrue(headTokenRel != null, TOKEN_NOT_ASSOCIATED_TO_ACCOUNT);
+
+            // Recreate the current head token's tokenRel, but with its previous pointer set to
+            // the last of the new tokenRels. This links the new token rels to the rest of the
+            // token rels connected via the old head token rel
+            final var lastOfNewTokenRels = newTokenRels.remove(newTokenRels.size() - 1);
+            final var headTokenAsNonHeadTokenRel = headTokenRel
+                    .copyBuilder()
+                    .previousToken(lastOfNewTokenRels.tokenNumber())
+                    .build(); // the old head token rel is no longer the head
+            // Also connect the last of the new tokenRels to the old head token rel
+            newTokenRels.add(lastOfNewTokenRels
+                    .copyBuilder()
+                    .nextToken(headTokenAsNonHeadTokenRel.tokenNumber())
+                    .build());
+            tokenRelStore.put(headTokenAsNonHeadTokenRel);
+        }
+
+        // Now replace the account's old head token number with the new head token number. This is
+        // how we link the new tokenRels to the account
+        final var firstOfNewTokenRels = newTokenRels.get(0);
+        final var updatedAcct = account.copyBuilder()
+                // replace the head token number with the first token number of the new tokenRels
+                .headTokenNumber(firstOfNewTokenRels.tokenNumber())
+                // and also update the account's total number of token associations
+                .numberAssociations(account.numberAssociations() + newTokenRels.size())
+                .build();
+
+        // Save the results
+        accountStore.put(updatedAcct);
+        newTokenRels.forEach(tokenRelStore::put);
+    }
+
+    /**
+     * Performs checks independent of state or context
+     */
+    private void pureChecks(@NonNull final TokenAssociateTransactionBody op) throws PreCheckException {
+        if (!op.hasAccount()) {
+            throw new PreCheckException(ResponseCodeEnum.INVALID_ACCOUNT_ID);
+        }
+
+        if (TokenListChecks.repeatsItself(op.tokensOrThrow())) {
+            throw new PreCheckException(TOKEN_ID_REPEATED_IN_TOKEN_LIST);
+        }
+    }
+
+    private record Validated(Account account, List<Token> tokens) {}
+
+    /**
+     * Performs checks that require state and context
+     */
+    private Validated validateSemantics(
+            @NonNull final List<TokenID> tokenIds,
+            @NonNull final AccountID accountId,
+            @NonNull final Configuration config,
+            @NonNull final WritableAccountStore accountStore,
+            @NonNull final ReadableTokenStore tokenStore,
+            @NonNull final WritableTokenRelationStore tokenRelStore) {
+        requireNonNull(config);
+
+        // Check that the system hasn't reached its limit of token associations
+        var maxTokenRels = config.getValue("maxNumTokenRels", Long.class);
```

Comment:
Wasn't sure if we're going to use the existing config names for this property and `maxTokensPerAccount` further down...?

Reply:
Sure, is there an example I can look at somewhere?

Reply:
I have added it in `TokenFeeSCheduleUpdate` PR